pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -846,61 +847,155 @@ func (m *kubeGenericRuntimeManager) purgeInitContainers(ctx context.Context, pod
 	}
 }
 
-// findNextInitContainerToRun returns the status of the last failed container, the
-// index of next init container to start, or done if there are no further init containers.
-// Status is only returned if an init container is failed, in which case next will
-// point to the current container.
-func findNextInitContainerToRun(pod *v1.Pod, podStatus *kubecontainer.PodStatus) (status *kubecontainer.Status, next *v1.Container, done bool) {
+// computeInitContainerActions sets the actions on the given changes that need
+// to be taken for the init containers. This includes actions to initialize the
+// init containers and actions to keep sidecar containers running.
+// computeInitContainerActions returns true if pod has been initialized.
+//
+// The actions include:
+// - Start the first init container that has not been started.
+// - Restart all sidecar containers that have started but are not running.
+// - Kill the sidecar containers that have failed the startup probe.
+func (m *kubeGenericRuntimeManager) computeInitContainerActions(pod *v1.Pod, podStatus *kubecontainer.PodStatus, changes *podActions) bool {
 	if len(pod.Spec.InitContainers) == 0 {
-		return nil, nil, true
+		return true
 	}
 
+	podHasInitialized := false
 	// If any of the main containers have status and are Running, then all init containers must
 	// have been executed at some point in the past.  However, they could have been removed
 	// from the container runtime now, and if we proceed, it would appear as if they
-	// never ran and will re-execute improperly.
+	// never ran and will re-execute improperly except for the sidecar containers.
 	for i := range pod.Spec.Containers {
 		container := &pod.Spec.Containers[i]
 		status := podStatus.FindContainerStatusByName(container.Name)
 		if status != nil && status.State == kubecontainer.ContainerStateRunning {
-			return nil, nil, true
+			podHasInitialized = true
+			break
 		}
 	}
 
-	// If there are failed containers, return the status of the last failed one.
-	for i := len(pod.Spec.InitContainers) - 1; i >= 0; i-- {
+	previousContainerInitialized := true
```

Comment:
comment to simplify the review:

This variable keeps track of whether or not the previous container was initialized.

Following this variable - it's clear that the logic WITHOUT sidecars is the same as it was before - search for non-running and non-failed regular init container.